design/comment-scanning.md

```diff
@@ -0,0 +1,110 @@
+- Documentation (README, docs/ entry, docs site entry, etc) that communicates these new definitions.
+
+## Non-Goals
+- Implement any general-purpose metadata/directive scanners. This document will focus solely on the syntax and the way we talk about scannable documentation going forward.
+
+
+## Background
+Currently, our analyzer is able to use documentation to detect implicit features and characteristics of features that would otherwise be missed. For example, the `@customElement` tag is used to signify that the following class definition is a custom element.
+
+However, there is another class of documentation that our analyzer currently doesn’t support. Our linter needs to be able to understand documentation that doesn’t describe a single, already existing feature. For example, the user may want to send an instruction to some analyzer-powered tool in its build chain:
+
+```html
+<!-- @registerServiceWorkerHere -->
+```
+
+Or, a user may want to describe some behavior across an entire source range, including any number of other existing features (including zero). For example:
+
+```js
+/* polymer-lint disable: undefined-elements */
+customElements.define('vanilla-element1', AnElementThatActuallyExists);
+customElements.define('vanilla-element2', IPromiseThisExists);
+customElements.define('vanilla-element3', SeriouslyLinterDontBeMad);
+/* polymer-lint enable: undefined-elements */
+```
+
+The polymer-lint enable/disable use-case is the immediate need here. But instead of building one-off support for this into the analyzer, we’d like to also think about how the analyzer should analyze all comment-based information so that future features don’t need to reinvent the wheel when it comes to comment-based analysis.
+
+
+## Design Overview
+
+This design groups all analysis-aware documentation into two supported formats: **Comment Directives** & **Comment Annotations**. Each supports a different general use of documentation, and together they cover most use cases for sending instructions and explicit information to the analyzer.
+
+### Comment Directive
+
+```
+/* polymer-lint disable: undefined-elements, rule-2 */
+/* polymer-lint enable: undefined-elements */
+/* @lazyImport '../some-url/some-file.html */
+<!-- @registerServiceWorkerHere -->
+```
+
+- A comment directive represents a new **Directive** feature type within a document
+- A new scanner should be written and used for each supported comment directive
+- A directive can describe one of three things:
+  - **Single Line Directive:** A feature that exists on a single line.
+  - **Source Range Directive Start:** A feature that exists on a source range within a document, starting at the directive’s source location.
+  - **Source Range Directive End:** A feature that exists on a source range within a document, ending at the directive’s source location.
+- A comment directive has a unique ID and may take arguments
```

Comment:
Will explain this more as a part of the fleshing out of the `Directive` feature type